README.md
# HyperGAN 0.10.0-alpha1

[![CircleCI](https://circleci.com/gh/255BITS/HyperGAN/tree/master.svg?style=svg)](https://circleci.com/gh/255BITS/HyperGAN/tree/master)
[![Discord](https://img.shields.io/badge/discord-join%20chat-brightgreen.svg)](https://discord.gg/t4WWBPF)

A composable GAN API and CLI.  Built for developers, researchers, and artists.

HyperGAN is currently in open beta.

![Colorizer 0.9 1](https://s3.amazonaws.com/hypergan-apidocs/0.9.0-images/colorizer-2.gif)

_Logos generated with [examples/colorizer](#examples),  AlphaGAN, and the RandomWalk sampler_


# Table of contents

* [About](#about)
* [Showcase](#showcase)
* [Documentation](#documentation)
* [Changelog](#changelog)
* [Quick start](#quick-start)
  * [Minimum Requirements](#minimum-requirements)
  * [Create a new project](#create-a-new-project)
  * [Install](#install)
  * [Train](#train)
  * [Increasing Performance](#increasing-performance)
  * [Development Mode](#development-mode)
  * [Running on CPU](#running-on-cpu)
* [The pip package hypergan](#the-pip-package-hypergan)
 * [Training](#training)
 * [Sampling](#sampling)
* [API](API)
  * [Examples](#examples)
  * [Search](#search)
* [Configuration](#configuration)
  * [Usage](#usage)
  * [Architecture](#architecture)
  * [GANComponent](#GANComponent)
  * [Generator](#generator)
  * [Encoders](#encoders)
  * [Discriminators](#discriminators)
  * [Losses](#losses)
   * [WGAN](#wgan)
   * [LS-GAN](#ls-gan)
   * [Standard GAN and Improved GAN](#standard-gan-and-improved-gan)
   * [Categories](#categorical-loss)
   * [Supervised](#supervised-loss)
  * [Trainers](#trainers)
* [Datasets](#datasets)
  * [Unsupervised learning](#unsupervised-learning)
  * [Supervised learning](#supervised-learning)
  * [Creating a Dataset](#creating-a-dataset)
  * [Downloadable Datasets](#downloadable-datasets)
* [Contributing](#contributing)
* [Versioning](#Versioning)
* [Sources](#sources)
* [Papers](#papers)
* [Citation](#citation)

# About

Generative Adversarial Networks consist of 2 learning systems that learn together.  HyperGAN implements these learning systems in Tensorflow with deep learning, using json files to reproduce architectures from white papers.

For an introduction, see here [http://blog.aylien.com/introduction-generative-adversarial-networks-code-tensorflow/](http://blog.aylien.com/introduction-generative-adversarial-networks-code-tensorflow/)


HyperGAN is a community project.  GANs are a very new and active field of research.  Join the community [discord](https://discord.gg/t4WWBPF).

## Features

* 100% community project
* Reproducible architectures using json configurations
* Domain Specific Language to define custom architectures
* API
* Builds graphs that can run on consumer hardware, like phones and web browsers
* Trainable on consumer hardware
* Dataset agnostic
* Combine different components to build your own GAN
* Transfer learning
* Optimistic loading allows for network changes at train time


# Showcase

Coming... eventually

# Documentation

## API Documentation

 * [0.10.0](https://s3.amazonaws.com/hypergan-apidocs/0.10.0/index.html)
 * [0.9.x](https://s3.amazonaws.com/hypergan-apidocs/0.9.0/index.html)
 * [Test coverage](https://s3.amazonaws.com/hypergan-apidocs/0.10.0/coverage/index.html)

# Changelog

See the full changelog here:
[Changelog.md](Changelog.md)

# Quick start

## Minimum requirements

1. For 256x256, we recommend a GTX 1080 or better.  32x32 can be run on lower-end GPUs.
2. CPU training is _extremely_ slow.  Use a GPU if you can!
3. Python3


## Install


### Install hypergan:

```bash
  pip3 install hypergan --upgrade
```

### Optional `virtualenv`:

If you use virtualenv:

```bash
  virtualenv --system-site-packages -p python3 hypergan
  source hypergan/bin/activate
```
### Dependencies:

If installation fails try this.

```bash
  pip3 install numpy tensorflow-gpu hyperchamber pillow pygame
```

### Dependency help

If the above step fails see the dependency documentation:

* tensorflow - https://www.tensorflow.org/install/
* pygame  - http://www.pygame.org/wiki/GettingStarted


## Create a new project

```bash
  hypergan new mymodel
```

This will create a mymodel.json based off the default configuration.  You can change configuration templates with the `-c` flag.  

## List configuration templates

```bash
  hypergan new mymodel -l
```

See all configuration templates with `--list-templates` or `-l`.

## Train

```bash
  # Train a 32x32 gan with batch size 32 on a folder of folders of pngs, resizing images as necessary
  hypergan train folder/ -s 32x32x3 -f png -c mymodel --resize
```

### Increasing performance

On ubuntu `sudo apt-get install libgoogle-perftools4` and make sure to include this environment variable before training

```bash
  LD_PRELOAD="/usr/lib/libtcmalloc.so.4" hypergan train my_dataset
```

HyperGAN does not cache image data in memory. Images are loaded every time they're needed, so you can increase performance by pre-processing your inputs, especially by resampling large inputs to the output resolution. e.g. with ImageMagick:

```bash
  convert image1.jpg -resize '128x128^' -gravity Center -crop 128x128+0+0 image1.png
```

## Development mode

If you wish to modify hypergan

```bash
git clone https://github.com/255BITS/hypergan
cd hypergan
python3 setup.py develop
```


## Running on CPU

Make sure to include the following 2 arguments:

```bash
CUDA_VISIBLE_DEVICES= hypergan --device '/cpu:0'
```
Don't train on CPU!  It's too slow.


# The pip package hypergan

```bash
 hypergan -h
```

## Training

```bash
  # Train a 32x32 gan with batch size 32 on a folder of pngs
  hypergan train [folder] -s 32x32x3 -f png -b 32 --config [name]
```

## Sampling

```bash
  # Train a 256x256 gan with batch size 32 on a folder of pngs
  hypergan train [folder] -s 32x32x3 -f png -b 32 --config [name] --sampler static_batch --sample_every 5 --save_samples
```

By default hypergan will not save samples to disk.  To change this, use `--save_samples`.

One way a network learns:

[![Demo CountPages alpha](https://j.gifs.com/58KmzA.gif)](https://www.youtube.com/watch?v=tj3ZLNfcJFo&list=PLWW3WtkBA3MuSnAVS__D0FkENZzuTbHFg&index=1)


To create videos:

```bash
  ffmpeg -i samples/%06d.png -vcodec libx264 -crf 22 -threads 0 gan.mp4
```
## Arguments

To see a detailed list, run 
```bash
  hypergan -h
```

# API

See the API documentation at https://s3.amazonaws.com/hypergan-apidocs/0.9.0/index.html

```python3
  import hypergan as hg
```

Examples
--------

See the example documentation https://github.com/255BITS/HyperGAN/tree/master/examples

## Search

Each example is capable of random search.  You can search along any set of parameters, including loss functions, trainers, generators, etc.

# Datasets

To build a new network you need a dataset.  Your data should be structured like:

``` 
  [folder]/[directory]/*.png
```

## Creating a Dataset

Datasets in HyperGAN are meant to be simple to create.  Just use a folder of images.

```
 [folder]/*.png
```

For jpg(pass `-f jpg`)

# Configuration

Configuration in HyperGAN uses JSON files.  You can create a new config with the default template by running `hypergan new mymodel`.

You can see all templates with `hypergan new mymodel -l`.


## Architecture

A hypergan configuration contains all hyperparameters for reproducing the full GAN.

In the original DCGAN you will have one of the following components:

* Distribution(latent space)
* Generator
* Discriminator
* Loss
* Trainer


Other architectures may differ.  See the configuration templates.

## GANComponent

A base class for each of the component types listed below.

## Generator

A generator is responsible for projecting an encoding (sometimes called *z space*) to an output (normally an image).  A single GAN object from HyperGAN has one generator.

## Distributions

Sometimes referred to as the `z-space` representation or `latent space`.  In `dcgan` the 'distribution' is random uniform noise.

Can be thought of as input to the `generator`.


### Uniform Distribution

| attribute   | description | type
|:----------:|:------------:|:----:|
| z | The dimensions of random uniform noise inputs | int > 0
| min | Lower bound of the random uniform noise | int
| max | Upper bound of the random uniform noise | int > min
| projections | See more about projections below | [f(config, gan, net):net, ...]
| modes | If using modes, the number of modes to have per dimension | int > 0


### Projections

This distribution takes a random uniform value and outputs it as many possible types.  The primary idea is that you are able to query Z as a random uniform distribution, even if the gan is using a spherical representation.

Some projection types are listed below.

#### "identity" projection

<img src='https://raw.githubusercontent.com/255BITS/HyperGAN/master/doc/encoder-linear-linear.png'/>

#### "sphere" projection

<img src='https://raw.githubusercontent.com/255BITS/HyperGAN/master/doc/encoder-linear-sphere.png'/>

#### "gaussian" projection

<img src='https://raw.githubusercontent.com/255BITS/HyperGAN/master/doc/encoder-linear-gaussian.png'/>

#### "modal" projection

One of many

#### "binary" projection

On/Off


### Category Distribution

Uses categorical prior to choose 'one-of-many' options.


## Discriminators

A discriminator's main purpose(sometimes called a critic) is to separate out G from X, and to give the Generator
a useful error signal to learn from.


## DSL

Each component in the GAN can be specified with a flexible DSL inside the JSON file.

## Losses

## WGAN

Wasserstein Loss is simply:

```python
 d_loss = d_real - d_fake
 g_loss = d_fake
```

d_loss and g_loss can be reversed as well - just add a '-' sign.

## Least-Squares GAN

```python
 d_loss = (d_real-b)**2 - (d_fake-a)**2
 g_loss = (d_fake-c)**2
```

a, b, and c are all hyperparameters.

### Standard GAN and Improved GAN

Includes support for Improved GAN.  See `hypergan/losses/standard_gan_loss.py` for details.

### Boundary Equilibrium Loss

Use with the `AutoencoderDiscriminator`.

See the `began` configuration template.

### Loss configuration

| attribute   | description | type
|:----------:|:------------:|:----:|
| batch_norm | batch_norm_1, layer_norm_1, or None | f(batch_size, name)(net):net
| create | Called during graph creation | f(config, gan, net):net
| discriminator |  Set to restrict this loss to a single discriminator(defaults to all) | int >= 0 or None
| label_smooth | improved gan - Label smoothing. | float > 0
| labels | lsgan - A triplet of values containing (a,b,c) terms. | [a,b,c] floats
| reduce | Reduces the output before applying loss | f(net):net
| reverse | Reverses the loss terms, if applicable | boolean

## Trainers

Determined by the GAN implementation.  These variables are the same across all trainers.

### Consensus

Consensus trainers trains G and D at the same time.  Resize Conv is known to not work with this technique(PR welcome).

#### Configuration

| attribute   | description | type
|:----------:|:------------:|:----:|
| learn_rate | Learning rate for the generator | float >= 0
| beta1 | (adam) | float >= 0
| beta2 | (adam)  | float >= 0
| epsilon | (adam)  | float >= 0
| decay | (rmsprop)  | float >= 0
| momentum | (rmsprop)  | float >= 0


### Alternating

Original GAN training.  Locks generator weights while training the discriminator, and vice-versa.

#### Configuration

| attribute   | description | type
|:----------:|:------------:|:----:|
| g_learn_rate | Learning rate for the generator | float >= 0
| g_beta1 | (adam) | float >= 0
| g_beta2 | (adam)  | float >= 0
| g_epsilon | (adam)  | float >= 0
| g_decay | (rmsprop)  | float >= 0
| g_momentum | (rmsprop)  | float >= 0
| d_learn_rate | Learning rate for the discriminator | float >= 0
| d_beta1 | (adam) | float >= 0
| d_beta2 | (adam)  | float >= 0
| d_epsilon | (adam)  | float >= 0
| d_decay | (rmsprop)  | float >= 0
| d_momentum | (rmsprop)  | float >= 0
| clipped_gradients | If set, gradients will be clipped to this value. | float > 0 or None
| d_clipped_weights | If set, the discriminator will be clipped by value. |float > 0 or None


### Fitness

Only trains on good z candidates.

### Curriculum

Train on a schedule.

### Gang

An evolution based trainer that plays a subgame between multiple generators/discriminators.


## Downloadable datasets

* CelebA aligned faces http://mmlab.ie.cuhk.edu.hk/projects/CelebA.html
* MS Coco http://mscoco.org/
* ImageNet http://image-net.org/
* youtube-dl (see [examples/Readme.md](examples/Readme.md))

# Contributing

Contributions are welcome and appreciated!  We have many open issues in the *Issues* tab.

See <a href='CONTRIBUTING.md'>how to contribute.</a>

# Versioning

HyperGAN uses semantic versioning.  http://semver.org/

TLDR: *x.y.z*

* _x_ is incremented on stable public releases.
* _y_ is incremented on API breaking changes.  This includes configuration file changes and graph construction changes.
* _z_ is incremented on non-API breaking changes.  *z* changes will be able to reload a saved graph.

## Papers

* GAN - https://arxiv.org/abs/1406.2661
* DCGAN - https://arxiv.org/abs/1511.06434
* InfoGAN - https://arxiv.org/abs/1606.03657
* Improved GAN - https://arxiv.org/abs/1606.03498
* Adversarial Inference - https://arxiv.org/abs/1606.00704
* Energy-based Generative Adversarial Network - https://arxiv.org/abs/1609.03126
* Wasserstein GAN - https://arxiv.org/abs/1701.07875
* Least Squares GAN - https://arxiv.org/pdf/1611.04076v2.pdf
* Boundary Equilibrium GAN - https://arxiv.org/abs/1703.10717
* Self-Normalizing Neural Networks - https://arxiv.org/abs/1706.02515
* Variational Approaches for Auto-Encoding
Generative Adversarial Networks - https://arxiv.org/pdf/1706.04987.pdf
* CycleGAN - https://junyanz.github.io/CycleGAN/
* DiscoGAN - https://arxiv.org/pdf/1703.05192.pdf
* Softmax GAN - https://arxiv.org/abs/1704.06191
* The Cramer Distance as a Solution to Biased Wasserstein Gradients - https://arxiv.org/abs/1705.10743
* Improved Training of Wasserstein GANs - https://arxiv.org/abs/1704.00028
* More...

## Sources

* DCGAN - https://github.com/carpedm20/DCGAN-tensorflow
* InfoGAN - https://github.com/openai/InfoGAN
* Improved GAN - https://github.com/openai/improved-gan
* Hyperchamber - https://github.com/255bits/hyperchamber

# Citation

If you wish to cite this project, do so like this:

```
  HyperGAN Community
  HyperGAN, (2017-2018+), 
  GitHub repository, 
  https://github.com/255BITS/HyperGAN
```

HyperGAN comes with no warranty or support.


Changelog.md
## 0.10 ~ Community Project release

* HyperGAN 0.10 released as a community project
* Configurable network architectures using a simple DSL
* Lots of new regularizers and losses
* Evolution based GangTrainer
* Curriculum trainer with progressive growing
* Optimistic loading
* New next-frame example
* More...

## 0.9 ~ Refactorings and optimizations

* API Documentation - https://s3.amazonaws.com/hypergan-apidocs/0.9.0/index.html
* Prepackaged configurations
* New viewer front-end!
* Examples, including the ability to randomly search for good configurations

See more here https://github.com/255BITS/HyperGAN/pull/66

## 0.8 ~ Least Squares GAN and API examples

* Tensorflow 1.0 support
* New configuration format and refactored api.
* New loss based on least squared GAN.  See <a href="#Least-Squares-GAN">least squares GAN implementation</a>.
* API example `2d-test` - tests a trainer/encoder/loss combination against a known distribution.
* API example `2d-measure` - measure and report the above test by randomly combining options.
* Updated default configuration.
* More


## 0.7 ~ WGAN & API

* New loss `wgan`
* Initial Public API Release
* API example: `colorizer` - re-colorize an image!
* API example: `inpainter` - remove a section of an image and have your GAN repaint it
* API example: `super-resolution` - zoom in and enhance.  We've caught the bad guy!
* 4 *new* samplers.  `--sampler` flag.  Valid options are: `batch`,`progressive`,`static_batch`,`grid`. 

## 0.5 / 0.6

* pip package released

<img src='https://raw.githubusercontent.com/255BITS/HyperGAN/master/doc/face-manifold-0-5-6.png'/>
<img src='https://raw.githubusercontent.com/255BITS/HyperGAN/master/doc/face-manifold.png'/>

### 0.1-0.4
* Initial private release

<img src='https://raw.githubusercontent.com/255BITS/HyperGAN/master/doc/legacy-0.1.png'/>
<img src='https://raw.githubusercontent.com/255BITS/HyperGAN/master/doc/legacy-0.1-2.png'/>





hypergan.py
import hyperchamber as hc
from lib.util.ops import *
from lib.util.globals import *
from lib.gan import *
from lib.util.gan_server import *
from tensorflow.contrib import ffmpeg
import lib.util.hc_tf as hc_tf
import lib.generators.resize_conv as resize_conv
import lib.generators.resize_conv_extra_layer as resize_conv_extra_layer
import lib.trainers.adam_trainer as adam_trainer
import lib.trainers.rmsprop_trainer as rmsprop_trainer
import lib.trainers.slowdown_trainer as slowdown_trainer
import lib.trainers.sgd_adam_trainer as sgd_adam_trainer
import lib.discriminators.pyramid_discriminator as pyramid_discriminator
import lib.discriminators.pyramid_nostride_discriminator as pyramid_nostride_discriminator
import lib.discriminators.slim_stride as slim_stride
import lib.discriminators.densenet_discriminator as densenet_discriminator
import lib.discriminators.fast_densenet_discriminator as fast_densenet_discriminator
import lib.discriminators.painters_discriminator as painters_discriminator
import lib.encoders.random_encoder as random_encoder
import lib.encoders.random_gaussian_encoder as random_gaussian_encoder
import lib.encoders.random_combo_encoder as random_combo_encoder
import lib.encoders.progressive_variational_encoder as progressive_variational_encoder
import lib.samplers.progressive_enhancement_sampler as progressive_enhancement_sampler
#import lib.samplers.grid_sampler as grid_sampler
import lib.regularizers.minibatch_regularizer as minibatch_regularizer
import lib.regularizers.moment_regularizer as moment_regularizer
import lib.regularizers.progressive_enhancement_minibatch_regularizer as progressive_enhancement_minibatch_regularizer
import lib.regularizers.l2_regularizer as l2_regularizer
import json
import uuid
import time

import lib.loaders.image_loader
import lib.loaders.audio_loader
import os
import sys
import time
import numpy as np
import tensorflow
import tensorflow as tf
import copy

import matplotlib
import matplotlib.pyplot as plt

from tensorflow.python.framework import ops

import importlib

import lib.cli as cli

args = cli.parse_args()

# Below are sets of configuration options:
# Each time a new random network is started a random set of configuration variables are selected.
# This is useful for hyperparameter search.  If you want to use a specific configuration use --config

hc.set('dtype', tf.float32) #The data type to use in our GAN.  Only float32 is supported at the moment

# Generator configuration
hc.set("generator", [resize_conv.generator, resize_conv_extra_layer.generator])
hc.set("generator.z", 100) # the size of the encoding.  Encoder is set by the 'encoder' property, but could just be a random_uniform
hc.set("generator.z_projection_depth", 1024) # Used in the first layer - the linear projection of z
hc.set("generator.activation", [prelu("g_")]); # activation function used inside the generator
hc.set("generator.activation.end", [tf.nn.tanh]); # Last layer of G.  Should match the range of your input - typically -1 to 1
hc.set("generator.fully_connected_layers", 0) # Experimental - This should probably stay 0
hc.set("generator.final_activation", [tf.nn.tanh]) #This should match the range of your input
hc.set("generator.resize_conv.depth_reduction", 2) # Divides our depth by this amount every time we go up in size
hc.set("generator.regularizers", [[l2_regularizer.get]]) # These are added to the loss function for G.
hc.set('generator.layer.noise', True) #Adds incremental noise each layer
hc.set("generator.regularizers.l2.lambda", list(np.linspace(0.1, 1, num=30))) # the magnitude of the l2 regularizer(experimental)
hc.set("generator.regularizers.layer", [layer_norm_1, batch_norm_1]) # the magnitude of the l2 regularizer(experimental)

# Trainer configuration
trainer = adam_trainer # adam works well at 64x64 but doesn't scale
#trainer = slowdown_trainer # this works at higher resolutions, but is slow and quirky(help wanted)
#trainer = rmsprop_trainer # this works at higher resolutions, but is slow and quirky(help wanted)
#trainer = sgd_adam_trainer # This has never worked, but seems like it should
hc.set("trainer.initializer", trainer.initialize) # TODO: can we merge these variables?
hc.set("trainer.train", trainer.train) # The training method to use.  This is called every step
hc.set("trainer.adam.discriminator.lr", 1e-3) #adam_trainer d learning rate
hc.set("trainer.adam.discriminator.epsilon", 1e-8) #adam epsilon for d
hc.set("trainer.adam.discriminator.beta1", 0.9) #adam beta1 for d
hc.set("trainer.adam.discriminator.beta2", 0.999) #adam beta2 for d
hc.set("trainer.adam.generator.lr", 1e-3) #adam_trainer g learning rate
hc.set("trainer.adam.generator.epsilon", 1e-8) #adam_trainer g
hc.set("trainer.adam.generator.beta1", 0.9) #adam_trainer g
hc.set("trainer.adam.generator.beta2", 0.999) #adam_trainer g
hc.set("trainer.rmsprop.discriminator.lr", 3e-5) # d learning rate
hc.set("trainer.rmsprop.generator.lr", 1e-4) # d learning rate
hc.set('trainer.slowdown.discriminator.d_fake_min', [0.12]) # healthy above this number on d_fake
hc.set('trainer.slowdown.discriminator.d_fake_max', [0.12001]) # unhealthy below this number on d_fake
hc.set('trainer.slowdown.discriminator.slowdown', [5]) # Divides speed by this number when unhealthy(d_fake low)
hc.set("trainer.sgd_adam.discriminator.lr", 3e-4) # d learning rate
hc.set("trainer.sgd_adam.generator.lr", 1e-3) # g learning rate

# Discriminator configuration
hc.set("discriminator", pyramid_nostride_discriminator.discriminator)
hc.set("discriminator.activation", [lrelu])#prelu("d_")])
hc.set('discriminator.regularizers.layer', layer_norm_1) # Size of fully connected layers

hc.set('discriminator.fc_layer', [False]) #If true, include a fully connected layer at the end of the discriminator
hc.set('discriminator.fc_layers', [0])# Number of fully connected layers to include
hc.set('discriminator.fc_layer.size', 378) # Size of fully connected layers

hc.set("discriminator.pyramid.layers", 5) #Layers in D
hc.set("discriminator.pyramid.depth_increase", 2)# Size increase of D's features on each layer

hc.set('discriminator.painters.layers', 2) #TODO has this ever worked?
hc.set('discriminator.painters.transitions', 5)
hc.set('discriminator.painters.activation', lrelu)

hc.set('discriminator.densenet.k', 32) #k is the number of features that are appended on each conv pass
hc.set('discriminator.densenet.layers', 1) #number of times to conv before size transition
hc.set('discriminator.densenet.transitions', 8) #number of transitions

hc.set('discriminator.add_noise', [True]) #add noise to input
hc.set('discriminator.noise_stddev', [1e-1]) #the amount of noise to add - always centered at 0
hc.set('discriminator.regularizers', [[minibatch_regularizer.get_features]]) # these regularizers get applied at the end of D

hc.set("sampler", progressive_enhancement_sampler.sample) # this is our sampling method.  Some other sampling ideas include cosine distance or adverarial encoding(not implemented but contributions welcome).
hc.set("sampler.samples", 3) # number of samples to generate at the end of each epoch
#hc.set('encoder', random_encoder.encode) # how to encode z
hc.set('encoder', random_combo_encoder.encode) # how to encode z

#hc.set("g_mp3_dilations",[[1,2,4,8,16,32,64,128,256]])
#hc.set("g_mp3_filter",[3])
#hc.set("g_mp3_residual_channels", [8])
#hc.set("g_mp3_dilation_channels", [16])
#hc.set("mp3_seconds", args.seconds)
#hc.set("mp3_bitrate", args.bitrate)
#hc.set("mp3_size", args.seconds*args.bitrate)

#TODO vae
#hc.set("transfer_fct", [tf.nn.elu, tf.nn.relu, tf.nn.relu6, lrelu]);
#hc.set("e_activation", [tf.nn.elu, tf.nn.relu, tf.nn.relu6, lrelu]);
#hc.set("e_last_layer", [tf.nn.tanh]);
#hc.set('f_skip_fc', False)
#hc.set('f_hidden_1', 512)#list(np.arange(256, 512)))
#hc.set('f_hidden_2', 256)#list(np.arange(256, 512)))

#TODO audio
hc.set("model", "faces:1.0")

hc.set("examples_per_epoch", 30000/4)

#TODO category/bernouilli
categories = [[]]#[[2]+[2]+build_categories_config(30)]
hc.set('categories', categories)
hc.set('categories_lambda', list(np.linspace(.001, .01, num=100)))
hc.set('category_loss', [False])

#TODO loss functions
hc.set('g_class_loss', [False])
hc.set('g_class_lambda', list(np.linspace(0.01, .1, num=30)))
hc.set('d_fake_class_loss', [False])

#TODO one-sided label smoothing loss
hc.set("g_target_prob", list(np.linspace(.65 /2., .85 /2., num=100)))
hc.set("d_label_smooth", list(np.linspace(0.15, 0.35, num=100)))

#TODO move to minibatch
hc.set("d_kernels", list(np.arange(10, 20)))
hc.set("d_kernel_dims", list(np.arange(100, 200)))

#TODO remove and replace with losses
hc.set("loss", ['custom'])

#TODO Vae Loss
hc.set("adv_loss", [False])
hc.set("latent_loss", [False])
hc.set("latent_lambda", list(np.linspace(.01, .1, num=30)))

#TODO Is this about adding z to the D?  Is this right? Investigate
hc.set("d_project", ['tiled'])

hc.set("batch_size", args.batch_size)

batch_no = 0

def frame_sample(sample_file, sess, config):
    """ Samples every frame to a file.  Useful for visualizing the learning process.

    Use with:

         ffmpeg -i samples/grid-%06d.png -vcodec libx264 -crf 22 -threads 0 grid1-7.mp4

    to create a video of the learning process.
    """

    if(args.frame_sample == None):
        return None
    if(args.frame_sample == "grid"):
        frame_sampler = grid_sampler.sample
    else:
        raise "Cannot find frame sampler: '"+args.frame_sample+"'"

    frame_sampler(sample_file, sess, config)

def epoch(sess, config):
    batch_size = config["batch_size"]
    n_samples =  config['examples_per_epoch']
    total_batch = int(n_samples / batch_size)
    global batch_no
    for i in range(total_batch):
        sample_file="samples/grid-%06d.png" % (batch_no * total_batch + i)
        frame_sample(sample_file, sess, config)

        d_loss, g_loss = config['trainer.train'](sess, config)

        #if(i > 10):
        #    if(math.isnan(d_loss) or math.isnan(g_loss) or g_loss > 1000 or d_loss > 1000):
        #        return False

        #    g = get_tensor('g')
        #    rX = sess.run([g[-1]])
        #    rX = np.array(rX)
        #    if(np.min(rX) < -1000 or np.max(rX) > 1000):
        #        return False
    batch_no+=1
    return True

def test_config(sess, config):
    batch_size = config["batch_size"]
    n_samples =  batch_size*10
    total_batch = int(n_samples / batch_size)
    results = []
    for i in range(total_batch):
        results.append(test(sess, config))
    return results

def collect_measurements(epoch, sess, config, time):
    d_loss = get_tensor("d_loss")
    d_loss_fake = get_tensor("d_fake_sig")
    d_loss_real = get_tensor("d_real_sig")
    g_loss = get_tensor("g_loss")
    d_class_loss = get_tensor("d_class_loss")
    simple_g_loss = get_tensor("g_loss_sig")

    gl, dl, dlr, dlf, dcl,sgl = sess.run([g_loss, d_loss, d_loss_real, d_loss_fake, d_class_loss, simple_g_loss])
    return {
            "g_loss": gl,
            "g_loss_sig": sgl,
            "d_loss": dl,
            "d_loss_real": dlr,
            "d_loss_fake": dlf,
            "d_class_loss": dcl,
            "g_strength": (1-(dlr))*(1-sgl),
            "seconds": time/1000.0
            }

def test_epoch(epoch, sess, config, start_time, end_time):
    sample = []
    sample_list = config['sampler'](sess,config)
    measurements = collect_measurements(epoch, sess, config, end_time - start_time)
    if args.use_hc_io:
        hc.io.measure(config, measurements)
        hc.io.sample(config, sample_list)
    else:
        print("Offline sample created:", sample_list)


def get_function(name):
    print(name)
    if name == "function:lib.util.ops.prelu_internal":
        return prelu("g_")

    if not isinstance(name, str):
        return name
    namespaced_method = name.split(":")[1]
    method = namespaced_method.split(".")[-1]
    namespace = ".".join(namespaced_method.split(".")[0:-1])
    return getattr(importlib.import_module(namespace),method)

# Take a config and replace any string starting with 'function:' with a function lookup.
def lookup_functions(config):
    for key, value in config.items():
        if(isinstance(value, str) and value.startswith("function:")):
            config[key]=get_function(value)
        if(isinstance(value, list) and len(value) > 0 and isinstance(value[0],str) and value[0].startswith("function:")):
            config[key]=[get_function(v) for v in value]
            
    return config


def output_graph_size():
    def mul(s):
        x = 1
        for y in s:
            x*=y
        return x
    def get_size(v):
        shape = [int(x) for x in v.get_shape()]
        size = mul(shape)
        return [v.name, size/1024./1024.]

    sizes = [get_size(i) for i in tf.all_variables()]
    sizes = sorted(sizes, key=lambda s: s[1])
    print("Top 5 sizes", sizes[-5:])
    size = sum([s[1] for s in sizes])
    print("SIZE = ", size)

def run(args):
    crop = args.crop
    channels = int(args.size.split("x")[2])
    width = int(args.size.split("x")[0])
    height = int(args.size.split("x")[1])
    loadedFromSave = False
    build_file = "build/generator.ckpt"

    print("Generating configs with hyper search space of ", hc.count_configs())
    for config in hc.configs(1):
        other_config = copy.copy(config)
        # load_saved_checkpoint(config)
        if(args.config):
            print("Loading config", args.config)

            if args.use_hc_io:
                config.update(hc.io.load_config(args.config))
            else:
                config = hc.load_or_create_config('~/.hypergan/configs/'+args.config+'.json', config)

        config = lookup_functions(config)
        config['batch_size']=args.batch_size

        config['dtype']=other_config['dtype']#TODO: add this as a CLI argument, i.e "-e 'dtype=function:tf.float16'"
        config['trainer.rmsprop.discriminator.lr']=other_config['trainer.rmsprop.discriminator.lr']

        # Initialize tensorflow
        with tf.device(args.device):
            sess = tf.Session(config=tf.ConfigProto())

        with tf.device('/cpu:0'):
            #TODO don't branch on format
            if(args.format == 'mp3'):
                x,y, num_labels,examples_per_epoch = lib.loaders.audio_loader.mp3_tensors_from_directory(args.directory,config['batch_size'], seconds=args.seconds, channels=channels, bitrate=args.bitrate, format=args.format)
                f = None
            else:
                x,y, f, num_labels,examples_per_epoch = lib.loaders.image_loader.labelled_image_tensors_from_directory(args.directory,config['batch_size'], channels=channels, format=args.format,crop=crop,width=width,height=height)

        config['y_dims']=num_labels
        config['x_dims']=[height,width] #TODO can we remove this?
        config['channels']=channels

        if args.config is None:
            filename = '~/.hypergan/configs/'+config['uuid']+'.json'
            print("Saving network configuration to: " + filename)
            config = hc.load_or_create_config(filename, config)

        with tf.device(args.device):
            y=tf.one_hot(tf.cast(y,tf.int64), config['y_dims'], 1.0, 0.0)

            if(args.method == 'build' or args.method == 'serve'):
                graph = create_generator(config,x,y,f)
            else:
                graph = create(config,x,y,f)

        #TODO can we not do this?  might need to be after hc.io refactor
        if('parent_uuid' in config):
            save_file = "saves/"+config["parent_uuid"]+".ckpt"
        else:
            save_file = "saves/"+config["uuid"]+".ckpt"

        print( "Save file", save_file,"\n")
        #TODO refactor save/load system
        if args.method == 'serve':
            print("|= Loading generator from build/")
            saver = tf.train.Saver()
            saver.restore(sess, build_file)
        elif(save_file and os.path.isfile(save_file+".index")):
            print(" |= Loading network from "+ save_file)
            config['uuid']=config['parent_uuid']
            ckpt = tf.train.get_checkpoint_state('saves')
            if ckpt and ckpt.model_checkpoint_path:
                saver = tf.train.Saver()
                saver.restore(sess, save_file)
                loadedFromSave = True
                print("Model loaded")
            else:
                print("No checkpoint file found")
        else:
            print(" |= Initializing new network")
            with tf.device(args.device):
                init = tf.initialize_all_variables()
                sess.run(init)

        output_graph_size()
        tf.train.start_queue_runners(sess=sess)
        testx = sess.run(x)

        if args.method == 'build':
            saver = tf.train.Saver()
            saver.save(sess, build_file)
            print("Saved generator to ", build_file)
        elif args.method == 'serve':
            gan_server(sess, config)
        else:
            sampled=False
            print("Running for ", args.epochs, " epochs")
            for i in range(args.epochs):
                start_time = time.time()
                with tf.device(args.device):
                    if(not epoch(sess, config)):
                        print("Epoch failed")
                        break
                print("Checking save "+ str(i))
                if(args.save_every != 0 and i % args.save_every == args.save_every-1):
                    print(" |= Saving network")
                    saver = tf.train.Saver()
                    saver.save(sess, save_file)
                end_time = time.time()
                test_epoch(i, sess, config, start_time, end_time)

            tf.reset_default_graph()
            sess.close()

if __name__ == "__main__":
    run(args) 
    

